src/main/scala/shine/OpenCL/CodeGeneration/InsertMemoryBarriers.scala

```diff
@@ -0,0 +1,208 @@
+                              // work-group parallel writes to outer scope allocations
+                              wg_writes: mutable.Map[Identifier[_ <: PhraseType], AddressSpace])
+
+  private def visitNewMetadata(p: Phrase[CommType],
+                               // allocations in the current scope
+                               allocs: Map[Identifier[_ <: PhraseType], AddressSpace]
+                              ): (Phrase[CommType], Metadata) = {
+    val meta = Metadata(mutable.Map(), mutable.Map())
+    val p2 = VisitAndRebuild(p, Visitor(allocs, meta))
+    (p2, meta)
+  }
+
+  private def visitLoopBody(
+    p: Phrase[CommType],
+    allocs: Map[Identifier[_ <: PhraseType], AddressSpace],
+    metadata: Metadata,
+    outer_wg_writes: mutable.Map[Identifier[_ <: PhraseType], AddressSpace] = mutable.Map()
+  ): Phrase[CommType] = {
+    val p2 = VisitAndRebuild(p, Visitor(allocs, metadata))
+    val dependencies = metadata.wg_writes.filter(kv => metadata.reads.contains(kv._1))
+    if (dependencies.nonEmpty) {
+      metadata.reads.clear()
+      metadata.wg_writes.clear()
+      Seq(p2, makeBarrier(dependencies.toMap))
+    } else {
+      metadata.wg_writes ++= outer_wg_writes
+      p2
+    }
+  }
+
+  private def makeBarrier(
+    allocs: Map[Identifier[_ <: PhraseType], AddressSpace]
+  ): Phrase[CommType] = {
+    OpenCL.DSL.barrier(
+      local = allocs.exists(_._2 == AddressSpace.Local),
+      global = allocs.exists(_._2 == AddressSpace.Global))
+  }
+
+  private case class Visitor(allocs: Map[Identifier[_ <: PhraseType], AddressSpace],
+                             metadata: Metadata) extends VisitAndRebuild.Visitor {
+    override def phrase[T <: PhraseType](p: Phrase[T]): Result[Phrase[T]] = {
+      p match {
+        case For(n, Lambda(x, body), unroll) =>
+          Stop(For(n, Lambda(x, visitLoopBody(body, allocs, metadata)), unroll))
+        case ForNat(n, DepLambda(x, body), unroll) =>
+          Stop(ForNat(n, DepLambda(x, visitLoopBody(body, allocs, metadata)), unroll))
+        case pf @ OpenCLParFor(n, dt, out, Lambda(x, Lambda(o, body)), init, step, unroll) =>
+          pf match {
+            case ParForLocal(dim) =>
+              val outer_wg_writes = mutable.Map[Identifier[_ <: PhraseType], AddressSpace]()
+              collectWrites(out, allocs, outer_wg_writes)
+              val b2 = visitLoopBody(body, allocs, metadata, outer_wg_writes)
+              Stop(ParForLocal(dim)(n, dt, out, Lambda(x, Lambda(o, b2)), init, step, unroll))
+            case ParForWorkGroup(dim) =>
+              Stop(ParForWorkGroup(dim)(n, dt, out,
+                Lambda(x, Lambda(o, visitLoopBody(body, allocs, metadata))), init, step, unroll))
+            case ParForGlobal(dim) =>
+              Stop(ParForGlobal(dim)(n, dt, out,
+                Lambda(x, Lambda(o, visitLoopBody(body, allocs, metadata))), init, step, unroll))
+          }
+        case pfn: OpenCLParForNat => ???
```

Comment:
Is it unclear what should happen here? Or is this just not copied over from the `OpenCLParFor` case?

Reply:
I did not investigate this pattern because it is not actively used in our current tests, but it should mostly be the same as the `OpenCLParFor` case.